### PR TITLE
Correctly capture the type parameter context symbol when generating SymbolId

### PR DESCRIPTION
**Pull request by Pilchie**
Fixes internal issue number 235912 (aka https://connect.microsoft.com/VisualStudio/feedback/details/2873787/stackoverflow-with-intellisense).

### CODE REVIEW

**Comment by Pilchie** [comment]
test vsi please

**Comment by jmarolf** [comment]
Do we have a testcase we can add for this?

**Comment by mattwar** [comment]
:+1:

**Comment by Pilchie** [comment]
@jmarolf indeed, and done.  Wanted to push before leaving yesterday so @mattwar could have a chance to see it.

**Comment by Pilchie** [comment]
test vsi please

**Comment by Pilchie** [comment]
@CyrusNajmabadi if you have access to the intertubes, I'd appreciate some :eyes:  on this...

**Comment by AlgorithmsAreCool** [comment]
@Pilchie Do you know if this fix would also work for recursive type constraints which seem to be causing similar issues? 

If so, should there be a test for those too in this PR?

**Comment by Pilchie** [comment]
@AlgorithmsAreCool - will spend some time trying to construct a repro of a constraint based issue if you haven't isolated it yet.

**Comment by AlgorithmsAreCool** [comment]
I'll keep trying to get a minimal repro also

**Comment by AlgorithmsAreCool** [comment]
@Pilchie Nevermind, I got a repro but only using a nested generic type. Your existing test covers it. I didn't notice that the crash was happening for me when i declared a property that was using a nested type.

Great fix :+1:

**Comment by Pilchie** [comment]
Updated to not regress partially constructed extension methods.  Another round of :eyes: appreciated.

**Comment by Pilchie** [comment]
test vsi please

**Comment by Pilchie** [comment]
@paulvanbrenk You expressed interest in taking a look at this today at shiproom to see if it has any impact on TypeScript.

File: src/Workspaces/Core/Portable/SymbolId/SymbolId.cs [hunk 1]
```diff
@@ -483,7 +483,16 @@ private void EncodeGenericTypeInfo(INamedTypeSymbol symbol)
                                 _builder.Append(",");
                             }
 
-                            this.Visit(symbol.TypeArguments[i]);
+                            // If we already have a type parameter context (say because we're encoding parameters of a generic method), then use it.
+                            // Otherwise, use the current type as the context, so that it's type parameters resolve
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/SymbolId/SymbolId.cs]: no apostrophe in "it's"

**Comment by CyrusNajmabadi** [comment]
:+1:

**Comment by paulvanbrenk** [comment]
As expected, this doesn't affect TypeScript. So LGTM

File: src/Workspaces/CoreTest/SymbolIdTests.cs [hunk 2]
```diff
@@ -427,6 +429,27 @@ public class C<S, T>
...
+}";
+            var compilation = GetCompilation(source);
+            var tree = compilation.SyntaxTrees.First();
+            var model = compilation.GetSemanticModel(tree);
+
+            var outer = GetDeclaredSymbols(compilation).OfType<INamedTypeSymbol>().First(s => s.Name == "A");
+            var constructed = outer.Construct(compilation.GetSpecialType(SpecialType.System_String));
+            var inner = constructed.GetTypeMembers().Single();
+            var id = SymbolId.CreateId(inner);
+            Assert.Equal("T:A{N:System.T:String}.T:B{`1}", id);
```
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/SymbolIdTests.cs]: Should this resolve the symbol back and then assert we get the same symbol?

File: src/Workspaces/CoreTest/SymbolIdTests.cs [hunk 3]
```diff
@@ -427,6 +429,27 @@ public class C<S, T>
...
+    public class B<TInner>
+    {
+    }
+}";
+            var compilation = GetCompilation(source);
+            var tree = compilation.SyntaxTrees.First();
+            var model = compilation.GetSemanticModel(tree);
+
+            var outer = GetDeclaredSymbols(compilation).OfType<INamedTypeSymbol>().First(s => s.Name == "A");
+            var constructed = outer.Construct(compilation.GetSpecialType(SpecialType.System_String));
```
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/SymbolIdTests.cs]: What about if the outer is unconstructed?

**Comment by jasonmalinowski** [comment]
No matter what a good set of tests around
```c#
class A<T1>
{
    class B<T2>
    {
        void M<T3>()
        {
        }
    }
}
```

seems needed, for all potential 2<sup>3</sup> scenarios where each type parameter is or isn't constructed. And that should verify round tripping, not just that the ID comes out in some form on the other side. :-(

**Comment by Pilchie** [comment]
Thanks @jasonmalinowski.  You've convinced me that this fix is not sufficient.  I have some tests that fail with the fix now.  I need to think about the right way to solve this.

**Comment by Pilchie** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
:+1:

**Comment by Pilchie** [comment]
FYI to all that this was released today, see https://msdn.microsoft.com/library/mt752379 for details.

**Comment by myvsissue** [comment]
Hi, 
  I am having same issue with Visual studio 2015 Update 3. As per the last post, I installed the update form msdn. However, that still did not fix my issue completely, it did improve it a bit though. Any help would be appreciated. I am seeing GetPrivateContextPerfCounters messages in stacktrace now and intellisense is still slow.

**Comment by Pilchie** [comment]
Hi @myvsissue, it sounds like you're seeing a different issue.  Can you report a feedback item through the VS titlebar and mention "@Pilchie" in it and let me know when you're done?  We'll take a look and see if we can figure out what's up.

**Comment by myvsissue** [comment]
I submitted a feedback item via VS toolbar...can you help as soon as possible?

**Pull request merged**